### PR TITLE
Test older model snapshots with ownership

### LINKED ISSUE #14445: After upgrade from 2.1 to 2.2, creating a new migration adds unexpected DropForeignKey, DropUniqueConstraint, DropColumn statements
### Issue summary
We have been using EF Core for 6 months now, and recently, after updating our project from 2.1 to 2.2, any new migration script has a ton of unexpected `DropForeignKey`, `DropUniqueConstraint`, `DropColumn` statements. Neither of the Keys, UniqueConstraints or Columns exists, so running this on the DB obviously fails. But this prevents us from creating new migrations.

I cannot come up with a small reproducible example, but I will try to share some patterns of the statements I noticed.

### Issue details
All DROP COLUMN statements happen only for classes that are owned by parent classes that are owned by a grandparent (it has to be double owning, single not enough). By owning I mean something like that:
```
builder.OwnsOne(
                nav => nav.MyCustomChildProperty, action =>
                {
                    action.OwnsOne(nav => nav.MyCustomGrandChildProperty);
                }
            );
```
In all such scenarions, in my migration scripts, I would see DROP COLUMN statements like such:
- `MyCustomChildProperty_TempId`
- `MyCustomChildProperty_TempId1` (this one is a maybe) 
- ...
- `MyCustomChildProperty_MyCustomGrandChildProperty_TempId`
- `MyCustomChildProperty_MyCustomGrandChildProperty_TempId1`  (this one is a maybe) 
- `MyCustomChildProperty_MyCustomGrandChildProperty_TempId2`  (this one is a maybe) 
- ...
and even weirder ones like:
- `MyCustomChildProperty_MyCustomGrandChildProperty_MyCustomChildPropertyClassMyCustomClassId`
Basically the pattern here is take class name of grandparent, add it to a grandparent class name, etc... and after reaching end, add `Id` to it.

I haven't tried to find a pattern for the other two types of statements (DropForeignKey, DropUniqueConstraint), but this seems that it should be related.

I tried deleting all of my migrations and creating a single fresh migration, and everything looks there as expected, therefore it seems that this is some bug that happens only if you already have some old migration scripts and trying to add a new one.

I have raised [an SO question](https://stackoverflow.com/questions/53819567/ef-core-migration-add-creates-a-migration-with-unexpected-dropforeignkey-drop) for this as well with no success.

### Possibly relevant versions
We were running using `2.1 versions of `dotnet` core and `ef core`.
Currently, my versions are:
- `dotnet --version` = `2.2.102`
- `dotnet ef --version` = `2.2.1-servicing-10028`
- Database Provider: `Microsoft.EntityFrameworkCore.SqlServer`, `v2.2.1`
- OS: Windows 10

### PR DESCRIPTION
**Pull request by bricelam**
Fixes #14445

### CODE REVIEW

File: src/EFCore.Design/Migrations/Internal/SnapshotModelProcessor.cs [hunk 1]
```diff
@@ -62,27 +63,52 @@ public virtual IModel Process(IModel model)
...
+            if ((version.StartsWith("2.0", StringComparison.Ordinal)
+                    || version.StartsWith("2.1", StringComparison.Ordinal))
+                && foreignKey.IsOwnership
+                && !foreignKey.PrincipalKey.IsPrimaryKey()
+                && foreignKey is IMutableForeignKey mutableForeignKey)
             {
-                ProcessElement(element, version);
+                mutableForeignKey.SetProperties(
+                    (IReadOnlyList<Property>)mutableForeignKey.Properties,
+                    mutableForeignKey.PrincipalEntityType.FindPrimaryKey());
```
Comment (AndriySvyryd) [inline src/EFCore.Design/Migrations/Internal/SnapshotModelProcessor.cs]: For this to work reliably with nested owned types we need to process the types in the topological order:

1. Start with all non-owned types
1. Go through all the types they own to set the principal key and the primary key
1. Repeat 2. recursively

Comment (bricelam) [reply src/EFCore.Design/Migrations/Internal/SnapshotModelProcessor.cs]: That sure cleaned up the implementation!

File: src/EFCore.Design/Migrations/Internal/SnapshotModelProcessor.cs [hunk 2]
```diff
@@ -135,5 +133,45 @@ private void ProcessElement(IAnnotatable metadata, string version)
...
+                var ownedType = ownership.DeclaringEntityType;
+
+                var oldPrincipalKey = ownership.PrincipalKey;
+                if (!oldPrincipalKey.IsPrimaryKey())
+                {
+                    ownership.SetProperties(
+                        (IReadOnlyList<Property>)ownership.Properties,
+                        ownership.PrincipalEntityType.FindPrimaryKey());
+
+                    oldPrincipalKey.DeclaringEntityType.RemoveKey(oldPrincipalKey);
```
Comment (AndriySvyryd) [inline src/EFCore.Design/Migrations/Internal/SnapshotModelProcessor.cs]: You should still check that the old key and properties were configured by convention

**Comment by wtgodbe** [comment]
Does this need to be in preview3, or is it for GA?

**Comment by bricelam** [comment]
The sooner the better, but either is fine

**Comment by AndriySvyryd** [comment]
...as long as it doesn't need extra approval

**Comment by ajcvickers** [comment]
@wtgodbe @Pilchie is following up with Steve on this.

**Pull request merged**